middlewares/middlewares.go

```diff
@@ -0,0 +1,106 @@
+package middlewares
+
+import (
+	"time"
+
+	"github.com/Clarilab/zoholab"
+	"github.com/go-resty/resty/v2"
+	"github.com/pkg/errors"
+)
+
+const (
+	accountsUri = "https://accounts.zoho.eu/oauth/v2/token"
+	timeout     = 3300
+)
+
+// ZohoService is the struct for the zoho service.
+type AuthTokenMiddleware struct {
+	clientID     string
+	clientSecret string
+	accessToken  *AuthToken
+	refreshToken string
+	lastRequest  *time.Time
+}
+
+// NewZohoService instantiates a new zoho service.
+func NewAuthTokenMiddleware(clientID, clientSecret, refreshtoken string) *AuthTokenMiddleware {
+	return &AuthTokenMiddleware{
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		refreshToken: refreshtoken,
+	}
+}
+
+// AddAuthToken adds an oauth token to a resty request.
+func (a *AuthTokenMiddleware) AddAuthTokenToRequest(client *resty.Client, request *resty.Request) error {
+	const errMsg = "could not add auth token"
+
+	token, err := a.getOAuthToken()
+	if err != nil {
+		return errors.Wrap(err, errMsg)
+	}
+
+	request.SetHeader("Authorization", "Zoho-oauthtoken "+token.AccessToken)
+
+	return nil
+}
+
+// getOAuthToken gets a oauthtoken from the zoho api.
+func (a *AuthTokenMiddleware) getOAuthToken() (*AuthToken, error) {
+	const errMsg = "could not get oauth token"
+
+	if a.isAccesTokenValid() {
+		return a.accessToken, nil
+	}
+
+	var result AuthToken
+
+	now := time.Now()
+
+	a.lastRequest = &now
+
+	resp, err := resty.New().R().
+		SetQueryParams(map[string]string{
+			"client_id":     a.clientID,
+			"client_secret": a.clientSecret,
+			"refresh_token": a.refreshToken,
+			"grant_type":    "refresh_token"},
+		).
+		SetResult(&result).
+		Post(accountsUri)
+	if err != nil {
+		return nil, errors.Wrap(err, errMsg)
+	}
+
+	if resp.IsError() {
+		return nil, errors.Wrap(zoholab.FillApiError(resp.Body()), errMsg)
+	}
+
+	return &result, nil
+}
+
+// isAccesTokenValid checks if an acces token is valid.
+func (a *AuthTokenMiddleware) isAccesTokenValid() bool {
+	if a.accessToken == nil || a.accessToken.AccessToken == "" {
+		return false
+	}
+
+	if a.lastRequest == nil || a.lastRequest.IsZero() {
+		return false
+	}
+
+	if time.Since(*a.lastRequest).Seconds() > float64(a.accessToken.ExpiresIn-timeout) {
```

Comment:
wieso ist das expiresIn-Timeout überhaupt notwendig? expiresIn sagt dir doch ziemlich genau, wann du wieder authentifizieren musst, wieso wird das hier nochmal manipuliert?

Reply:
Das expiresIn ist immer eine Stunde und wird auch nicht irgendwie geupdated oder andersweitig geändert deswegen setzte ich dieses LastRequest demenprechend, damit ich weis wann der letzte request gefeuert wurde und dann hab ich mir einen timeout überlegt der mir sinnvoll vorkam. Soll heißen das wenn der Last request länger als 55min her war wir eine neuer accesstoken geholt.

Reply:
Ja genau aber das ist doch immer eine Stunde ab dem Zeitpunkt wo du das Token das letzte mal requested hast
Heißt, sobald time.Since(lastRequest) > expiresIn ist, willst du nen neuen Token ziehen, oder nicht?

Reply:
Ja das ist ja genau der check ich hab nur den timeout eingeführt das da nicht irgendwie eine Sekunde übrig ist und der request deswegen failed. Meinst du der ist nicht notwendig?

Reply:
aber wird hier nicht effektiv nach 5 min neu authentifiziert? 😄 expiresIn == 60min - timeout== 55min

Reply:
ne das muss ja kleiner sein als time since last request diese 55min hab dafür auch einen kleinen test geschrieben

Reply:
dein test checkt aber nur 1min und 1h, was passiert denn bei 10min z.B.? Wie gesagt, ich denke das wird auch neu authentifizieren wollen, was irgendwie unnötig ist